File: java/RemoveNodesFromLinkedList2487.java
/*
You are given the head of a linked list.
Remove every node which has a node with a greater value anywhere to the right side of it.
Return the head of the modified linked list.
*/
public class RemoveNodesFromLinkedList2487 {
    public static void main(String[] args) {
        ListNode head = new ListNode(4);
        ListNode node1 = new ListNode(5);
        head.next = node1;
        ListNode node2 = new ListNode(1);
        node1.next = node2;
        ListNode node3 = new ListNode(9);
        node2.next = node3;
        ListNode node = head;
        while (node != null) {
            System.out.print(node.val + " ");
            node = node.next;
        }
        System.out.println();
        head = removeNodes(head);
        node = head;
        while (node != null) {
            System.out.print(node.val + " ");
            node = node.next;
        }
        System.out.println();
        head = new ListNode(1);
        node1 = new ListNode(2);
        head.next = node1;
        node2 = new ListNode(3);
        node1.next = node2;
        node3 = new ListNode(4);
        node2.next = node3;
        node3.next = new ListNode(5);
        node = head;
        while (node != null) {
            System.out.print(node.val + " ");
            node = node.next;
        }
        System.out.println();
        head = removeNodes(head);
        node = head;
        while (node != null) {
            System.out.print(node.val + " ");
            node = node.next;
        }
    }

    public static ListNode removeNodes(ListNode head) {
        head = reverse(head);
        ListNode current = head;
        int currentMax = current.val;
        while (current.next != null) {
            if (current.next.val < currentMax) current.next = current.next.next;
            else {
                current = current.next;
                currentMax = current.val;
            }
        }
        return reverse(head);
    }

    private static ListNode reverse(ListNode head) {
        ListNode previous = null;
        while (head != null) {
            ListNode nextHead = head.next;
            head.next = previous;
            previous = head;
            head = nextHead;
        }
        return previous;
    }

    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}

File: python/RemoveNodesFromLinkedList2487.py
# You are given the head of a linked list.
# Remove every node which has a node with a greater value anywhere to the right side of it.
# Return the head of the modified linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def removeNodes(head: Optional[ListNode]) -> Optional[ListNode]:
    def reverse(node):
        previous = None
        while node:
            node.next, node, previous = previous, node.next, node
        return previous

    head = reverse(head)
    current = head
    current_max = current.val
    while current.next:
        if current.next.val < current_max:
            current.next = current.next.next
        else:
            current = current.next
            current_max = current.val
    return reverse(head)


if __name__ == "__main__":
    head_outer = ListNode(1)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(4)
    head_outer.next.next.next.next = ListNode(5)
    head_outer = removeNodes(head_outer)
    while head_outer:
        print(head_outer.val, end=' ')
        head_outer = head_outer.next
    print()
    head_outer = ListNode(5)
    head_outer.next = ListNode(2)
    head_outer.next.next = ListNode(3)
    head_outer.next.next.next = ListNode(8)
    head_outer.next.next.next.next = ListNode(6)
    head_outer.next.next.next.next.next = ListNode(1)
    head_outer.next.next.next.next.next.next = ListNode(7)
    head_outer = removeNodes(head_outer)
    while head_outer:
        print(head_outer.val, end=' ')
        head_outer = head_outer.next
    print()
    head_outer = ListNode(4)
    head_outer.next = ListNode(5)
    head_outer.next.next = ListNode(2)
    head_outer.next.next.next = ListNode(3)
    head_outer.next.next.next.next = ListNode(8)
    head_outer.next.next.next.next.next = ListNode(6)
    head_outer.next.next.next.next.next.next = ListNode(1)
    head_outer.next.next.next.next.next.next.next = ListNode(7)
    head_outer = removeNodes(head_outer)
    while head_outer:
        print(head_outer.val, end=' ')
        head_outer = head_outer.next
    print()
    head_outer = ListNode(4)
    head_outer.next = ListNode(5)
    head_outer.next.next = ListNode(2)
    head_outer.next.next.next = ListNode(3)
    head_outer.next.next.next.next = ListNode(8)
    head_outer.next.next.next.next.next = ListNode(6)
    head_outer.next.next.next.next.next.next = ListNode(1)
    head_outer.next.next.next.next.next.next.next = ListNode(7)
    head_outer = removeNodes(head_outer)
    while head_outer:
        print(head_outer.val, end=' ')
        head_outer = head_outer.next
    print()
    head_outer = ListNode(4)
    head_outer.next = ListNode(5)
    head_outer.next.next = ListNode(2)
    head_outer.next.next.next = ListNode(3)
    head_outer.next.next.next.next = ListNode(8)
    head_outer.next.next.next.next.next = ListNode(6)
    head_outer.next.next.next.next.next.next = ListNode(1)
    head_outer.next.next.next.next.next.next.next = ListNode(7)
    head_outer = removeNodes(head_outer)
    while head_outer:
        print(head_outer.val, end=' ')
        head_outer = head_outer.next
